syntax = "proto3";

// =========================== USER SERVICE ==========================
message User {
  int32 id = 1;
  string username = 2; // Unique username
  string email = 3; // Email address, unique
  string phone = 4; // Phone number (optional)
  bool is_active = 5; // Indicates if the user is active
  bool is_verified = 6; // Indicates if the user has verified the account
  string role = 7; // Role (e.g., "Customer", "Admin")
  string created_at = 8; // ISO 8601 format for account creation timestamp
}

message UserCreate {
  int32 id = 1;
  string username = 2;
  string email = 3;
  string password_hash = 4;
  string phone = 5; // Phone number (optional)
  bool is_active = 6;
  bool is_verified = 7;
  string role = 8; // User role
  string created_at = 9; // Account creation timestamp
  string updated_at = 10;
}

message UserUpdate {
  int32 id = 1;
  string password_hash = 2;
  string phone = 3;
  string updated_at = 4; // ISO 8601 format for profile last update timestamp
}

message UserDelete {
  int32 id = 1;
}

// =========================== USER PROFILE ==========================
message UserProfile {
  int32 id = 1;
  string first_name = 2;
  string last_name = 3;
  string address = 4;
  string avatar_url = 5;
  string date_of_birth = 6; // ISO 8601 date format
  int32 user_id = 7; // Link to User
}

message UserProfileCreate {
  int32 id = 1;
  string first_name = 2;
  string last_name = 3;
  string address = 4;
  string avatar_url = 5;
  string date_of_birth = 6;
  int32 user_id = 7;
}

message UserProfileUpdate {
  int32 id = 1;
  string first_name = 2;
  string last_name = 3;
  string address = 4;
  string avatar_url = 5;
}

// ========================== AUTH & PASSWORD =======================
message UserLoginRequest {
  string username = 1;
  string password = 2;
}

message UserLoginResponse {
  int32 user_id = 1;
  string access_token = 2;
  string refresh_token = 3;
}

message PasswordResetRequest {
 string username=1;
 string oldpassword=2;
 string newpassword=3;

}

message PasswordResetConfirm {
  string reset_token = 1;
  string new_password = 2;
}

// ============================ USER EVENT ==========================
message UserEvent {
  enum EventType {
    CREATED = 0;
    UPDATED = 1;
    DELETED = 2;
    VERIFIED = 3;
  }

  int32 id = 1;
  int32 user_id = 2;
  EventType event_type = 3;
  string event_payload = 4; // JSON-encoded user data
  string timestamp = 5; // Event timestamp in ISO 8601 format
}

// ========================== USER LIST RESPONSE ====================
message UserList {
  repeated User users = 1;
}

message UserProfileList {
  repeated UserProfile profiles = 1;
}
